src/Std/Sat/AIG/If.lean

```diff
@@ -0,0 +1,280 @@
+  generalize hgo : go aig curr hcurr discr lhs rhs s = res
+  unfold go at hgo
+  dsimp at hgo
+  split at hgo
+  . rw [← hgo]
+    intro idx h1 h2
+    rw [go_decl_eq]
+    rw [AIG.LawfulOperator.decl_eq (f := AIG.mkIfCached)]
+    apply AIG.LawfulOperator.lt_size_of_lt_aig_size (f := AIG.mkIfCached)
+    assumption
+  . simp [← hgo]
+termination_by w - curr
+
+end ite
+
+instance : LawfulStreamOperator α IfInput ite where
+  le_size := by
+    intros
+    unfold ite
+    apply ite.go_le_size
+  decl_eq := by
+    intros
+    unfold ite
+    rw [ite.go_decl_eq]
+
+namespace ite
+
+theorem go_get_aux {w : Nat} (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w) (discr : Ref aig)
+    (lhs rhs : RefStream aig w) (s : RefStream aig curr) :
+    ∀ (idx : Nat) (hidx : idx < curr) (hfoo),
+      (go aig curr hcurr discr lhs rhs s).stream.get idx (by omega)
+        =
+      (s.get idx hidx).cast hfoo := by
+  intro idx hidx
+  generalize hgo : go aig curr hcurr discr lhs rhs s = res
+  unfold go at hgo
+  dsimp at hgo
+  split at hgo
+  . rw [← hgo]
+    intros
+    rw [go_get_aux]
+    rw [AIG.RefStream.get_push_ref_lt]
+    . simp only [Ref.cast, Ref.mk.injEq]
+      rw [AIG.RefStream.get_cast]
+      . simp
+      . assumption
+    . apply go_le_size
+  . rw [← hgo]
+    simp only [Nat.le_refl, get, Ref_cast', Ref.mk.injEq, true_implies]
+    have : curr = w := by omega
+    subst this
+    simp
+termination_by w - curr
+
+theorem go_get {w : Nat} (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w) (discr : Ref aig)
+    (lhs rhs : RefStream aig w) (s : RefStream aig curr) :
+    ∀ (idx : Nat) (hidx : idx < curr),
+      (go aig curr hcurr discr lhs rhs s).stream.get idx (by omega)
+        =
+      (s.get idx hidx).cast (by apply go_le_size) := by
+  intro idx hidx
+  apply go_get_aux
+
+theorem go_denote_mem_prefix {w : Nat} (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w)
+    (discr : Ref aig) (lhs rhs : RefStream aig w) (s : RefStream aig curr) (start : Nat) (hstart) :
+    ⟦
+      (go aig curr hcurr discr lhs rhs s).aig,
+      ⟨start, by apply Nat.lt_of_lt_of_le; exact hstart; apply go_le_size⟩,
+      assign
+    ⟧
+      =
+    ⟦aig, ⟨start, hstart⟩, assign⟧ := by
+  apply denote.eq_of_IsPrefix (entry := ⟨aig, start,hstart⟩)
+  apply IsPrefix.of
+  . intros
+    apply go_decl_eq
+  . intros
+    apply go_le_size
+
+theorem denote_go {w : Nat} (aig : AIG α) (curr : Nat) (hcurr : curr ≤ w) (discr : Ref aig)
+    (lhs rhs : RefStream aig w) (s : RefStream aig curr) :
+    ∀ (idx : Nat) (hidx1 : idx < w),
+      curr ≤ idx
+        →
+      ⟦
+        (go aig curr hcurr discr lhs rhs s).aig,
+        (go aig curr hcurr discr lhs rhs s).stream.get idx hidx1,
+        assign
+      ⟧
+        =
+      if ⟦aig, discr, assign⟧ then
+        ⟦aig, lhs.get idx hidx1, assign⟧
+      else
+        ⟦aig, rhs.get idx hidx1, assign⟧ := by
+  intro idx hidx1 hidx2
+  generalize hgo : go aig curr hcurr discr lhs rhs s = res
+  unfold go at hgo
+  dsimp at hgo
```

Comment:
Non-terminal simp?

Reply:
`dsimp` only applies definitional reductions if not provided with arguments as far as I know?. Its job here is to reduce the tons of lets that are in the term to a single let free term that I can rewrite on.

Reply:
I think `dsimp only at hgo` is the way to go here.